{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ef0e5946",
   "metadata": {},
   "source": [
    "(sec_intro)=\n",
    "\n",
    "# Population Genomics in Practice exercises\n",
    "\n",
    "Welcome to the Population Genomics in Practice exercises homepage!\n",
    "\n",
    "(sec_exercises_about)=\n",
    "\n",
    "## About the exercises\n",
    "\n",
    "As the focus of the course is on hands-on work, the topics have been\n",
    "designed to cover the fundamental analyses that are common in many\n",
    "population genomics studies. \n",
    "\n",
    "(sec_exercises_about_manuscript_route)=\n",
    "\n",
    "### The manuscript route\n",
    "\n",
    "In principle, you could imagine the course structure to follow that of\n",
    "a manuscript (e.g {cite}`fuller_PopulationGeneticsCoral_2020`).\n",
    "\n",
    "High-throughput DNA sequencing has now made it possible to generate\n",
    "whole-genome resequencing data for multiple individuals and\n",
    "populations, and a first step is to map sequence data to a reference,\n",
    "perform variant calling and variant filtering.\n",
    "\n",
    "Once a high-quality variant set has been obtained, a common task is to\n",
    "describe variation, either in terms of summary statistics such as\n",
    "nucleotide diversity ($\\pi$) or site-frequency spectra (sfs), or as\n",
    "descriptions of population structure in terms of admixture or pca\n",
    "plots.\n",
    "\n",
    "Genetic diversity is also affected by population history and\n",
    "demographic processes such as population expansion, bottlenecks,\n",
    "migration events and hybridizations.\n",
    "\n",
    "Finally, it is often of interest to identify adaptive traits, to which\n",
    "end selection tests and scans can be performed. The tests are designed\n",
    "to detect signals of selection, either via direct selection on loci,\n",
    "or by looking at haplotype structures to detect linked selection.\n",
    "\n",
    "(sec_exercises_about_baseline)=\n",
    "\n",
    "### The baseline model\n",
    "\n",
    "Much of what has been described in [The manuscript\n",
    "route](sec_exercises_about_manuscript_route) has recently been treated\n",
    "in an article on statistical inference in population genomics\n",
    "{cite}`johri_StatisticalInferencePopulation_2021`. In it, the authors\n",
    "point out that whereas historically theoretical advances outpaced data\n",
    "production, that is no longer true due to the advent of\n",
    "next-generation sequencing. In particular, they caution researchers to\n",
    "attach too much faith to a test that explains the data well, as there\n",
    "are many alternative hypotheses with equal explanatory power, but with\n",
    "drastically different conclusions. At the very least, a population\n",
    "genomics study should aim at first generating a baseline model\n",
    "consisting of all or several of the following components:\n",
    "\n",
    "1. mutation\n",
    "2. recombination\n",
    "3. reassortment\n",
    "4. gene conversion\n",
    "5. purifying selection acting on functional regions and its effects on\n",
    "   linked variants (background selection)\n",
    "6. genetic drift with demographic history and geographic structure\n",
    "\n",
    "The exercises are designed to address many of the points above, and to\n",
    "highlight cases where competing hypotheses may actually explain data\n",
    "to equal degrees.\n",
    "\n",
    "(sec_exercises_about_additional_material)=\n",
    "\n",
    "### Additional material\n",
    "\n",
    "```{admonition} FIXME\n",
    ":class: warning\n",
    "Rewrite/restructure and add more additional topics\n",
    "```\n",
    "\n",
    "\n",
    "```{note}\n",
    "NB: with time we could use repo to collect all kinds of exercises\n",
    "(even those that don't fit current curriculum), providing the\n",
    "possibility to create optional course plans\n",
    "``` \n",
    "\n",
    "A five day course cannot but provide an overview of possible analyses\n",
    "and topics. Among the things we won't have time to cover are\n",
    "\n",
    "1. machine learning and AI for population genomics\n",
    "2. recombination landscape estimation\n",
    "3. gene conversion\n",
    "4. experimental design\n",
    "5. coalescent and Wright-Fisher simulations\n",
    "6. ABC methods\n",
    "7. ancestral recombination graphs and tree sequence inference\n",
    "8. ancient DNA\n",
    "9. spatial genetics\n",
    "10. ..."
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "pgip"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "source_map": [
   12
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}