{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56f309e7",
   "metadata": {},
   "source": [
    "```{note}\n",
    "About: exercises on variant calling. Input data consists of simulated and/or experimental data.\n",
    "```\n",
    "\n",
    "(sec_variantcalling)=\n",
    "\n",
    "# Variant calling\n",
    "\n",
    "\n",
    "## Best practice pipeline\n",
    "\n",
    "Briefly describe best practice pipeline (e.g. GATK, bcftools,\n",
    "freebayes), highlighting some of the problems with natural\n",
    "populations:\n",
    "- many workflows are human-centric and don't scale to non-model\n",
    "  organisms (e.g. chromosome numbers in GATK)\n",
    "  \n",
    "\n",
    "## Variant calling exercise\n",
    "\n",
    "```{tabbed} bcftools\n",
    "```{code-block} shell\n",
    "bcftools | head -2\n",
    "```\n",
    "\n",
    "```{tabbed} gatk\n",
    "```{code-block} shell\n",
    "gatk HaplotypeCaller\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d389b01f",
   "metadata": {
    "tags": [
     "remove-input",
     "hide-output"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Program: bcftools (Tools for variant calling and manipulating VCFs and BCFs)\n",
      "License: GNU GPLv3+, due to use of the GNU Scientific Library\n",
      "Version: 1.15.1 (using htslib 1.15.1)\n",
      "\n",
      "Usage:   bcftools [--version|--version-only] [--help] <command> <argument>\n",
      "\n",
      "Commands:\n",
      "\n",
      " -- Indexing\n",
      "    index        index VCF/BCF files\n",
      "\n",
      " -- VCF/BCF manipulation\n",
      "    annotate     annotate and edit VCF/BCF files\n",
      "    concat       concatenate VCF/BCF files from the same set of samples\n",
      "    convert      convert VCF/BCF files to different formats and back\n",
      "    head         view VCF/BCF file headers\n",
      "    isec         intersections of VCF/BCF files\n",
      "    merge        merge VCF/BCF files files from non-overlapping sample sets\n",
      "    norm         left-align and normalize indels\n",
      "    plugin       user-defined plugins\n",
      "    query        transform VCF/BCF into user-defined formats\n",
      "    reheader     modify VCF/BCF header, change sample names\n",
      "    sort         sort VCF/BCF file\n",
      "    view         VCF/BCF conversion, view, subset and filter VCF/BCF files\n",
      "\n",
      " -- VCF/BCF analysis\n",
      "    call         SNP/indel calling\n",
      "    consensus    create consensus sequence by applying VCF variants\n",
      "    cnv          HMM CNV calling\n",
      "    csq          call variation consequences\n",
      "    filter       filter VCF/BCF files using fixed thresholds\n",
      "    gtcheck      check sample concordance, detect sample swaps and contamination\n",
      "    mpileup      multi-way pileup producing genotype likelihoods\n",
      "    polysomy     detect number of chromosomal copies\n",
      "    roh          identify runs of autozygosity (HMM)\n",
      "    stats        produce VCF/BCF stats\n",
      "\n",
      " -- Plugins (collection of programs for calling, file manipulation & analysis)\n",
      "    38 plugins available, run \"bcftools plugin -lv\" to see a complete list\n",
      "\n",
      " Most commands accept VCF, bgzipped VCF, and BCF with the file type detected\n",
      " automatically even when streaming from a pipe. Indexed VCF and BCF will work\n",
      " in all situations. Un-indexed VCF and BCF and streams will work in most but\n",
      " not all situations.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='bcftools | head -2', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "subprocess.run(\"bcftools | head -2\", shell=True, check=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dbc7d4c6",
   "metadata": {
    "tags": [
     "remove-input",
     "hide-output"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GATK jar /usr/share/miniconda/envs/test/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /usr/share/miniconda/envs/test/share/gatk4-4.2.6.1-1/gatk-package-4.2.6.1-local.jar HaplotypeCaller\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "USAGE: HaplotypeCaller [arguments]\n",
      "\n",
      "Call germline SNPs and indels via local re-assembly of haplotypes\n",
      "Version:4.2.6.1\n",
      "\n",
      "\n",
      "Required Arguments:\n",
      "\n",
      "--input,-I <GATKPath>         BAM/SAM/CRAM file containing reads  This argument must be specified at least once.\n",
      "                              Required. \n",
      "\n",
      "--output,-O <GATKPath>        File to which variants should be written  Required. \n",
      "\n",
      "--reference,-R <GATKPath>     Reference sequence file  Required. \n",
      "\n",
      "\n",
      "Optional Arguments:\n",
      "\n",
      "--add-output-sam-program-record <Boolean>\n",
      "                              If true, adds a PG tag to created SAM/BAM/CRAM files.  Default value: true. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--add-output-vcf-command-line <Boolean>\n",
      "                              If true, adds a command line header line to created VCF files.  Default value: true.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--alleles <FeatureInput>      The set of alleles to force-call regardless of evidence  Default value: null. \n",
      "\n",
      "--annotate-with-num-discovered-alleles <Boolean>\n",
      "                              If provided, we will annotate records with the number of alternate alleles that were\n",
      "                              discovered (but not necessarily genotyped) at a given site  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--annotation,-A <String>      One or more specific annotations to add to variant calls  This argument may be specified 0\n",
      "                              or more times. Default value: null. Possible values: {AlleleFraction, AllelePseudoDepth,\n",
      "                              AS_BaseQualityRankSumTest, AS_FisherStrand, AS_InbreedingCoeff,\n",
      "                              AS_MappingQualityRankSumTest, AS_QualByDepth, AS_ReadPosRankSumTest, AS_RMSMappingQuality,\n",
      "                              AS_StrandBiasMutectAnnotation, AS_StrandOddsRatio, AssemblyComplexity, BaseQuality,\n",
      "                              BaseQualityHistogram, BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest,\n",
      "                              CountNs, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet,\n",
      "                              FeaturizedReadSets, FisherStrand, FragmentDepthPerAlleleBySample, FragmentLength,\n",
      "                              GenotypeSummaries, InbreedingCoeff, LikelihoodRankSumTest, MappingQuality,\n",
      "                              MappingQualityRankSumTest, MappingQualityZero, OrientationBiasReadCounts,\n",
      "                              OriginalAlignment, PossibleDeNovo, QualByDepth, ReadPosition, ReadPosRankSumTest,\n",
      "                              ReferenceBases, RMSMappingQuality, SampleList, StrandBiasBySample, StrandOddsRatio,\n",
      "                              TandemRepeat, UniqueAltReadCountPossible values: {\n",
      "\n",
      "--annotation-group,-G <String>One or more groups of annotations to apply to variant calls  This argument may be\n",
      "                              specified 0 or more times. Default value: null. Possible values:\n",
      "                              {AlleleSpecificAnnotation, AS_StandardAnnotation, GenotypeAnnotation, InfoFieldAnnotation,\n",
      "                              JumboGenotypeAnnotation, JumboInfoAnnotation, ReducibleAnnotation, StandardAnnotation,\n",
      "                              StandardHCAnnotation, StandardMutectAnnotation, VariantAnnotationPossible values: {\n",
      "\n",
      "--annotations-to-exclude,-AX <String>\n",
      "                              One or more specific annotations to exclude from variant calls  This argument may be\n",
      "                              specified 0 or more times. Default value: null. Possible values: {BaseQualityRankSumTest,\n",
      "                              ChromosomeCounts, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, ExcessHet,\n",
      "                              FisherStrand, InbreedingCoeff, MappingQualityRankSumTest, QualByDepth, ReadPosRankSumTest,\n",
      "                              RMSMappingQuality, StrandOddsRatioPossible values: {\n",
      "\n",
      "--arguments_file <File>       read one or more arguments files and add them to the command line  This argument may be\n",
      "                              specified 0 or more times. Default value: null. \n",
      "\n",
      "--assembly-region-out <String>Output the assembly region to this IGV formatted file  Default value: null. \n",
      "\n",
      "--assembly-region-padding <Integer>\n",
      "                              Number of additional bases of context to include around each assembly region  Default\n",
      "                              value: 100. \n",
      "\n",
      "--base-quality-score-threshold <Byte>\n",
      "                              Base qualities below this threshold will be reduced to the minimum (6)  Default value: 18.\n",
      "\n",
      "--cloud-index-prefetch-buffer,-CIPB <Integer>\n",
      "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable). Defaults to\n",
      "                              cloudPrefetchBuffer if unset.  Default value: -1. \n",
      "\n",
      "--cloud-prefetch-buffer,-CPB <Integer>\n",
      "                              Size of the cloud-only prefetch buffer (in MB; 0 to disable).  Default value: 40. \n",
      "\n",
      "--contamination-fraction-to-filter,-contamination <Double>\n",
      "                              Fraction of contamination in sequencing data (for all samples) to aggressively remove \n",
      "                              Default value: 0.0. \n",
      "\n",
      "--create-output-bam-index,-OBI <Boolean>\n",
      "                              If true, create a BAM/CRAM index when writing a coordinate-sorted BAM/CRAM file.  Default\n",
      "                              value: true. Possible values: {true, false} \n",
      "\n",
      "--create-output-bam-md5,-OBM <Boolean>\n",
      "                              If true, create a MD5 digest for any BAM/SAM/CRAM file created  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--create-output-variant-index,-OVI <Boolean>\n",
      "                              If true, create a VCF index when writing a coordinate-sorted VCF file.  Default value:\n",
      "                              true. Possible values: {true, false} \n",
      "\n",
      "--create-output-variant-md5,-OVM <Boolean>\n",
      "                              If true, create a a MD5 digest any VCF file created.  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--dbsnp,-D <FeatureInput>     dbSNP file  Default value: null. \n",
      "\n",
      "--disable-bam-index-caching,-DBIC <Boolean>\n",
      "                              If true, don't cache bam indexes, this will reduce memory requirements but may harm\n",
      "                              performance if many intervals are specified.  Caching is automatically disabled if there\n",
      "                              are no intervals specified.  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-read-filter,-DF <String>\n",
      "                              Read filters to be disabled before analysis  This argument may be specified 0 or more\n",
      "                              times. Default value: null. Possible values: {GoodCigarReadFilter, MappedReadFilter,\n",
      "                              MappingQualityAvailableReadFilter, MappingQualityReadFilter,\n",
      "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
      "                              NotSecondaryAlignmentReadFilter, PassesVendorQualityCheckReadFilter,\n",
      "                              WellformedReadFilterPossible values: {\n",
      "\n",
      "--disable-sequence-dictionary-validation <Boolean>\n",
      "                              If specified, do not check the sequence dictionaries from our inputs for compatibility.\n",
      "                              Use at your own risk!  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-use-dragstr-pair-hmm-scores <Boolean>\n",
      "                              disable DRAGstr pair-hmm score even when dragstr-params-path was provided  Default value:\n",
      "                              false. Possible values: {true, false} \n",
      "\n",
      "--dragen-mode <Boolean>       Single argument for enabling the bulk of DRAGEN-GATK features. NOTE: THIS WILL OVERWRITE\n",
      "                              PROVIDED ARGUMENT CHECK TOOL INFO TO SEE WHICH ARGUMENTS ARE SET).  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--dragstr-het-hom-ratio <Integer>\n",
      "                              het to hom prior ratio use with DRAGstr on  Default value: 2. \n",
      "\n",
      "--dragstr-params-path <GATKPath>\n",
      "                              location of the DRAGstr model parameters for STR error correction used in the Pair HMM.\n",
      "                              When provided, it overrides other PCR error correcting mechanisms  Default value: null. \n",
      "\n",
      "--enable-dynamic-read-disqualification-for-genotyping <Boolean>\n",
      "                              Will enable less strict read disqualification low base quality reads  Default value:\n",
      "                              false. Possible values: {true, false} \n",
      "\n",
      "--exclude-intervals,-XL <String>\n",
      "                              One or more genomic intervals to exclude from processing  This argument may be specified 0\n",
      "                              or more times. Default value: null. \n",
      "\n",
      "--founder-id <String>         Samples representing the population \"founders\"  This argument may be specified 0 or more\n",
      "                              times. Default value: null. \n",
      "\n",
      "--gatk-config-file <String>   A configuration file to use with the GATK.  Default value: null. \n",
      "\n",
      "--gcs-max-retries,-gcs-retries <Integer>\n",
      "                              If the GCS bucket channel errors out, how many times it will attempt to re-initiate the\n",
      "                              connection  Default value: 20. \n",
      "\n",
      "--gcs-project-for-requester-pays <String>\n",
      "                              Project to bill when accessing \"requester pays\" buckets. If unset, these buckets cannot be\n",
      "                              accessed.  User must have storage.buckets.get permission on the bucket being accessed. \n",
      "                              Default value: . \n",
      "\n",
      "--genotype-assignment-method,-gam <GenotypeAssignmentMethod>\n",
      "                              How we assign genotypes  Default value: USE_PLS_TO_ASSIGN. Possible values:\n",
      "                              {SET_TO_NO_CALL, USE_PLS_TO_ASSIGN, USE_POSTERIORS_ANNOTATION,\n",
      "                              SET_TO_NO_CALL_NO_ANNOTATIONS, BEST_MATCH_TO_ORIGINAL, DO_NOT_ASSIGN_GENOTYPES,\n",
      "                              USE_POSTERIOR_PROBABILITIES, PREFER_PLS} \n",
      "\n",
      "--graph-output,-graph <String>Write debug assembly graph information to this file  Default value: null. \n",
      "\n",
      "--help,-h <Boolean>           display the help message  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--heterozygosity <Double>     Heterozygosity value used to compute prior probabilities for any locus.  See the GATKDocs\n",
      "                              for full details on the meaning of this population genetics concept  Default value: 0.001.\n",
      "\n",
      "--heterozygosity-stdev <Double>\n",
      "                              Standard deviation of heterozygosity for SNP and indel calling.  Default value: 0.01. \n",
      "\n",
      "--indel-heterozygosity <Double>\n",
      "                              Heterozygosity for indel calling.  See the GATKDocs for heterozygosity for full details on\n",
      "                              the meaning of this population genetics concept  Default value: 1.25E-4. \n",
      "\n",
      "--interval-exclusion-padding,-ixp <Integer>\n",
      "                              Amount of padding (in bp) to add to each interval you are excluding.  Default value: 0. \n",
      "\n",
      "--interval-merging-rule,-imr <IntervalMergingRule>\n",
      "                              Interval merging rule for abutting intervals  Default value: ALL. Possible values: {ALL,\n",
      "                              OVERLAPPING_ONLY} \n",
      "\n",
      "--interval-padding,-ip <Integer>\n",
      "                              Amount of padding (in bp) to add to each interval you are including.  Default value: 0. \n",
      "\n",
      "--interval-set-rule,-isr <IntervalSetRule>\n",
      "                              Set merging approach to use for combining interval inputs  Default value: UNION. Possible\n",
      "                              values: {UNION, INTERSECTION} \n",
      "\n",
      "--intervals,-L <String>       One or more genomic intervals over which to operate  This argument may be specified 0 or\n",
      "                              more times. Default value: null. \n",
      "\n",
      "--lenient,-LE <Boolean>       Lenient processing of VCF files  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--max-assembly-region-size <Integer>\n",
      "                              Maximum size of an assembly region  Default value: 300. \n",
      "\n",
      "--max-reads-per-alignment-start <Integer>\n",
      "                              Maximum number of reads to retain per alignment start position. Reads above this threshold\n",
      "                              will be downsampled. Set to 0 to disable.  Default value: 50. \n",
      "\n",
      "--max-variants-per-shard <Integer>\n",
      "                              If non-zero, partitions VCF output into shards, each containing up to the given number of\n",
      "                              records.  Default value: 0. \n",
      "\n",
      "--min-assembly-region-size <Integer>\n",
      "                              Minimum size of an assembly region  Default value: 50. \n",
      "\n",
      "--min-base-quality-score,-mbq <Byte>\n",
      "                              Minimum base quality required to consider a base for calling  Default value: 10. \n",
      "\n",
      "--native-pair-hmm-threads <Integer>\n",
      "                              How many threads should a native pairHMM implementation use  Default value: 4. \n",
      "\n",
      "--native-pair-hmm-use-double-precision <Boolean>\n",
      "                              use double precision in the native pairHmm. This is slower but matches the java\n",
      "                              implementation better  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--num-reference-samples-if-no-call <Integer>\n",
      "                              Number of hom-ref genotypes to infer at sites not present in a panel  Default value: 0. \n",
      "\n",
      "--output-mode <OutputMode>    Specifies which type of calls we should output  Default value: EMIT_VARIANTS_ONLY.\n",
      "                              Possible values: {EMIT_VARIANTS_ONLY, EMIT_ALL_CONFIDENT_SITES, EMIT_ALL_ACTIVE_SITES} \n",
      "\n",
      "--pedigree,-ped <GATKPath>    Pedigree file for determining the population \"founders\"  Default value: null. \n",
      "\n",
      "--population-callset,-population <FeatureInput>\n",
      "                              Callset to use in calculating genotype priors  Default value: null. \n",
      "\n",
      "--QUIET <Boolean>             Whether to suppress job-summary info on System.err.  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--read-filter,-RF <String>    Read filters to be applied before analysis  This argument may be specified 0 or more\n",
      "                              times. Default value: null. Possible values: {AlignmentAgreesWithHeaderReadFilter,\n",
      "                              AllowAllReadsReadFilter, AmbiguousBaseReadFilter, CigarContainsNoNOperator,\n",
      "                              ExcessiveEndClippedReadFilter, FirstOfPairReadFilter, FragmentLengthReadFilter,\n",
      "                              GoodCigarReadFilter, HasReadGroupReadFilter, IntervalOverlapReadFilter, LibraryReadFilter,\n",
      "                              MappedReadFilter, MappingQualityAvailableReadFilter, MappingQualityNotZeroReadFilter,\n",
      "                              MappingQualityReadFilter, MatchingBasesAndQualsReadFilter, MateDifferentStrandReadFilter,\n",
      "                              MateDistantReadFilter, MateOnSameContigOrNoMappedMateReadFilter,\n",
      "                              MateUnmappedAndUnmappedReadFilter, MetricsReadFilter,\n",
      "                              NonChimericOriginalAlignmentReadFilter, NonZeroFragmentLengthReadFilter,\n",
      "                              NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter,\n",
      "                              NotOpticalDuplicateReadFilter, NotProperlyPairedReadFilter,\n",
      "                              NotSecondaryAlignmentReadFilter, NotSupplementaryAlignmentReadFilter,\n",
      "                              OverclippedReadFilter, PairedReadFilter, PassesVendorQualityCheckReadFilter,\n",
      "                              PlatformReadFilter, PlatformUnitReadFilter, PrimaryLineReadFilter,\n",
      "                              ProperlyPairedReadFilter, ReadGroupBlackListReadFilter, ReadGroupReadFilter,\n",
      "                              ReadLengthEqualsCigarLengthReadFilter, ReadLengthReadFilter, ReadNameReadFilter,\n",
      "                              ReadStrandFilter, SampleReadFilter, SecondOfPairReadFilter, SeqIsStoredReadFilter,\n",
      "                              SoftClippedReadFilter, ValidAlignmentEndReadFilter, ValidAlignmentStartReadFilter,\n",
      "                              WellformedReadFilterPossible values: {\n",
      "\n",
      "--read-index <GATKPath>       Indices to use for the read inputs. If specified, an index must be provided for every read\n",
      "                              input and in the same order as the read inputs. If this argument is not specified, the\n",
      "                              path to the index for each input will be inferred automatically.  This argument may be\n",
      "                              specified 0 or more times. Default value: null. \n",
      "\n",
      "--read-validation-stringency,-VS <ValidationStringency>\n",
      "                              Validation stringency for all SAM/BAM/CRAM/SRA files read by this program.  The default\n",
      "                              stringency value SILENT can improve performance when processing a BAM file in which\n",
      "                              variable-length data (read, qualities, tags) do not otherwise need to be decoded.  Default\n",
      "                              value: SILENT. Possible values: {STRICT, LENIENT, SILENT} \n",
      "\n",
      "--recover-dangling-heads <Boolean>\n",
      "                              This argument is deprecated since version 3.3  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--sample-name,-ALIAS <String> Name of single sample to use from a multi-sample bam  Default value: null. \n",
      "\n",
      "--sample-ploidy,-ploidy <Integer>\n",
      "                              Ploidy (number of chromosomes) per sample. For pooled data, set to (Number of samples in\n",
      "                              each pool * Sample Ploidy).  Default value: 2. \n",
      "\n",
      "--seconds-between-progress-updates <Double>\n",
      "                              Output traversal statistics every time this many seconds elapse  Default value: 10.0. \n",
      "\n",
      "--sequence-dictionary <GATKPath>\n",
      "                              Use the given sequence dictionary as the master/canonical sequence dictionary.  Must be a\n",
      "                              .dict file.  Default value: null. \n",
      "\n",
      "--sites-only-vcf-output <Boolean>\n",
      "                              If true, don't emit genotype fields when writing vcf file output.  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--standard-min-confidence-threshold-for-calling,-stand-call-conf <Double>\n",
      "                              The minimum phred-scaled confidence threshold at which variants should be called  Default\n",
      "                              value: 30.0. \n",
      "\n",
      "--tmp-dir <GATKPath>          Temp directory to use.  Default value: null. \n",
      "\n",
      "--use-jdk-deflater,-jdk-deflater <Boolean>\n",
      "                              Whether to use the JdkDeflater (as opposed to IntelDeflater)  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--use-jdk-inflater,-jdk-inflater <Boolean>\n",
      "                              Whether to use the JdkInflater (as opposed to IntelInflater)  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--use-new-qual-calculator,-new-qual <Boolean>\n",
      "                              Use the new AF model instead of the so-called exact model  Default value: true. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--use-posteriors-to-calculate-qual,-gp-qual <Boolean>\n",
      "                              if available, use the genotype posterior probabilities to calculate the site QUAL  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "--verbosity <LogLevel>        Control verbosity of logging.  Default value: INFO. Possible values: {ERROR, WARNING,\n",
      "                              INFO, DEBUG} \n",
      "\n",
      "--version <Boolean>           display the version number for this tool  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "\n",
      "Advanced Arguments:\n",
      "\n",
      "--active-probability-threshold <Double>\n",
      "                              Minimum probability for a locus to be considered active.  Default value: 0.002. \n",
      "\n",
      "--adaptive-pruning <Boolean>  Use Mutect2's adaptive graph pruning algorithm  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--adaptive-pruning-initial-error-rate <Double>\n",
      "                              Initial base error rate estimate for adaptive pruning  Default value: 0.001. \n",
      "\n",
      "--all-site-pls <Boolean>      Annotate all sites with PLs  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--allele-informative-reads-overlap-margin <Integer>\n",
      "                              Likelihood and read-based annotations will only take into consideration reads that overlap\n",
      "                              the variant or any base no further than this distance expressed in base pairs  Default\n",
      "                              value: 2. \n",
      "\n",
      "--allow-non-unique-kmers-in-ref <Boolean>\n",
      "                              Allow graphs that have non-unique kmers in the reference  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--apply-bqd <Boolean>         If enabled this argument will apply the DRAGEN-GATK BaseQualityDropout model to the\n",
      "                              genotyping model for filtering sites due to Linked Error mode.  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--apply-frd <Boolean>         If enabled this argument will apply the DRAGEN-GATK ForeignReadDetection model to the\n",
      "                              genotyping model for filtering sites.  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--bam-output,-bamout <String> File to which assembled haplotypes should be written  Default value: null. \n",
      "\n",
      "--bam-writer-type <WriterType>Which haplotypes should be written to the BAM  Default value: CALLED_HAPLOTYPES. Possible\n",
      "                              values: {ALL_POSSIBLE_HAPLOTYPES, CALLED_HAPLOTYPES, NO_HAPLOTYPES} \n",
      "\n",
      "--comparison,-comp <FeatureInput>\n",
      "                              Comparison VCF file(s)  This argument may be specified 0 or more times. Default value:\n",
      "                              null. \n",
      "\n",
      "--contamination-fraction-per-sample-file,-contamination-file <File>\n",
      "                              Tab-separated File containing fraction of contamination in sequencing data (per sample) to\n",
      "                              aggressively remove. Format should be \"<SampleID><TAB><Contamination>\" (Contamination is\n",
      "                              double) per line; No header.  Default value: null. \n",
      "\n",
      "--debug-assembly,-debug <Boolean>\n",
      "                              Print out verbose debug information about each assembly region  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--disable-cap-base-qualities-to-map-quality <Boolean>\n",
      "                              If false this disables capping of base qualities in the HMM to the mapping quality of the\n",
      "                              read  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-optimizations <Boolean>\n",
      "                              Don't skip calculations in ActiveRegions with no variants  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--disable-spanning-event-genotyping <Boolean>\n",
      "                              If enabled this argument will disable inclusion of the '*' spanning event when genotyping\n",
      "                              events that overlap deletions  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-symmetric-hmm-normalizing <Boolean>\n",
      "                              Toggle to revive legacy behavior of asymmetrically normalizing the arguments to the\n",
      "                              reference haplotype  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--disable-tool-default-annotations <Boolean>\n",
      "                              Disable all tool default annotations  Default value: false. Possible values: {true, false}\n",
      "\n",
      "--disable-tool-default-read-filters <Boolean>\n",
      "                              Disable all tool default read filters (WARNING: many tools will not function correctly\n",
      "                              without their default read filters on)  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--do-not-correct-overlapping-quality <Boolean>\n",
      "                              Disable overlapping base quality correction  Default value: false. Possible values: {true,\n",
      "                              false} \n",
      "\n",
      "--do-not-run-physical-phasing <Boolean>\n",
      "                              Disable physical phasing  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-increase-kmer-sizes-for-cycles <Boolean>\n",
      "                              Disable iterating over kmer sizes when graph cycles are detected  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "--dont-use-dragstr-priors <Boolean>\n",
      "                              Forfeit the use of the DRAGstr model to calculate genotype priors. This argument does not\n",
      "                              have any effect in the absence of DRAGstr model parameters (--dragstr-model-params) \n",
      "                              Default value: false. Possible values: {true, false} \n",
      "\n",
      "--dont-use-soft-clipped-bases <Boolean>\n",
      "                              Do not analyze soft clipped bases in the reads  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--emit-ref-confidence,-ERC <ReferenceConfidenceMode>\n",
      "                              Mode for emitting reference confidence scores (For Mutect2, this is a BETA feature) \n",
      "                              Default value: NONE. Possible values: {NONE, BP_RESOLUTION, GVCF} \n",
      "\n",
      "--enable-all-annotations <Boolean>\n",
      "                              Use all possible annotations (not for the faint of heart)  Default value: false. Possible\n",
      "                              values: {true, false} \n",
      "\n",
      "--expected-mismatch-rate-for-read-disqualification <Double>\n",
      "                              Error rate used to set expectation for post HMM read disqualification based on mismatches \n",
      "                              Default value: 0.02. \n",
      "\n",
      "--floor-blocks <Boolean>      Output the band lower bound for each GQ block regardless of the data it represents \n",
      "                              Default value: false. Possible values: {true, false} \n",
      "\n",
      "--force-active <Boolean>      If provided, all regions will be marked as active  Default value: false. Possible values:\n",
      "                              {true, false} \n",
      "\n",
      "--force-call-filtered-alleles,-genotype-filtered-alleles <Boolean>\n",
      "                              Force-call filtered alleles included in the resource specified by --alleles  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "--gvcf-gq-bands,-GQB <Integer>Exclusive upper bounds for reference confidence GQ bands (must be in [1, 100] and\n",
      "                              specified in increasing order)  This argument may be specified 0 or more times. Default\n",
      "                              value: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,\n",
      "                              24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45,\n",
      "                              46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 70, 80, 90, 99]. \n",
      "\n",
      "--indel-size-to-eliminate-in-ref-model <Integer>\n",
      "                              The size of an indel to check for in the reference model  Default value: 10. \n",
      "\n",
      "--kmer-size <Integer>         Kmer size to use in the read threading assembler  This argument may be specified 0 or more\n",
      "                              times. Default value: [10, 25]. \n",
      "\n",
      "--linked-de-bruijn-graph <Boolean>\n",
      "                              If enabled, the Assembly Engine will construct a Linked De Bruijn graph to recover better\n",
      "                              haplotypes  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--mapping-quality-threshold-for-genotyping <Integer>\n",
      "                              Control the threshold for discounting reads from the genotyper due to mapping quality\n",
      "                              after the active region detection and assembly steps but before genotyping. NOTE: this is\n",
      "                              in contrast to the --minimum-mapping-quality argument which filters reads from all parts\n",
      "                              of the HaplotypeCaller. If you would like to call genotypes with a different threshold\n",
      "                              both arguments must be set.  Default value: 20. \n",
      "\n",
      "--max-alternate-alleles <Integer>\n",
      "                              Maximum number of alternate alleles to genotype  Default value: 6. \n",
      "\n",
      "--max-effective-depth-adjustment-for-frd <Integer>\n",
      "                              Set the maximum depth to modify FRD adjustment to in the event of high depth sites (0 to\n",
      "                              disable)  Default value: 0. \n",
      "\n",
      "--max-genotype-count <Integer>Maximum number of genotypes to consider at any site  Default value: 1024. \n",
      "\n",
      "--max-mnp-distance,-mnp-dist <Integer>\n",
      "                              Two or more phased substitutions separated by this distance or less are merged into MNPs. \n",
      "                              Default value: 0. \n",
      "\n",
      "--max-num-haplotypes-in-population <Integer>\n",
      "                              Maximum number of haplotypes to consider for your population  Default value: 128. \n",
      "\n",
      "--max-prob-propagation-distance <Integer>\n",
      "                              Upper limit on how many bases away probability mass can be moved around when calculating\n",
      "                              the boundaries between active and inactive assembly regions  Default value: 50. \n",
      "\n",
      "--max-unpruned-variants <Integer>\n",
      "                              Maximum number of variants in graph the adaptive pruner will allow  Default value: 100. \n",
      "\n",
      "--min-dangling-branch-length <Integer>\n",
      "                              Minimum length of a dangling branch to attempt recovery  Default value: 4. \n",
      "\n",
      "--min-pruning <Integer>       Minimum support to not prune paths in the graph  Default value: 2. \n",
      "\n",
      "--num-pruning-samples <Integer>\n",
      "                              Number of samples that must pass the minPruning threshold  Default value: 1. \n",
      "\n",
      "--pair-hmm-gap-continuation-penalty <Integer>\n",
      "                              Flat gap continuation penalty for use in the Pair HMM  Default value: 10. \n",
      "\n",
      "--pair-hmm-implementation,-pairHMM <Implementation>\n",
      "                              The PairHMM implementation to use for genotype likelihood calculations  Default value:\n",
      "                              FASTEST_AVAILABLE. Possible values: {EXACT, ORIGINAL, LOGLESS_CACHING,\n",
      "                              AVX_LOGLESS_CACHING, AVX_LOGLESS_CACHING_OMP, FASTEST_AVAILABLE} \n",
      "\n",
      "--pair-hmm-results-file <GATKPath>\n",
      "                              File to write exact pairHMM inputs/outputs to for debugging purposes  Default value: null.\n",
      "\n",
      "--pcr-indel-model <PCRErrorModel>\n",
      "                              The PCR indel model to use  Default value: CONSERVATIVE. Possible values: {NONE, HOSTILE,\n",
      "                              AGGRESSIVE, CONSERVATIVE} \n",
      "\n",
      "--phred-scaled-global-read-mismapping-rate <Integer>\n",
      "                              The global assumed mismapping rate for reads  Default value: 45. \n",
      "\n",
      "--pruning-lod-threshold <Double>\n",
      "                              Ln likelihood ratio threshold for adaptive pruning algorithm  Default value:\n",
      "                              2.302585092994046. \n",
      "\n",
      "--pruning-seeding-lod-threshold <Double>\n",
      "                              Ln likelihood ratio threshold for seeding subgraph of good variation in adaptive pruning\n",
      "                              algorithm  Default value: 9.210340371976184. \n",
      "\n",
      "--recover-all-dangling-branches <Boolean>\n",
      "                              Recover all dangling branches  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--showHidden <Boolean>        display hidden arguments  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--smith-waterman <Implementation>\n",
      "                              Which Smith-Waterman implementation to use, generally FASTEST_AVAILABLE is the right\n",
      "                              choice  Default value: JAVA. Possible values: {FASTEST_AVAILABLE, AVX_ENABLED, JAVA} \n",
      "\n",
      "--smith-waterman-dangling-end-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for dangling-end recovery.  Default value: -6. \n",
      "\n",
      "--smith-waterman-dangling-end-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for dangling-end recovery.  Default value: -110. \n",
      "\n",
      "--smith-waterman-dangling-end-match-value <Integer>\n",
      "                              Smith-Waterman match value for dangling-end recovery.  Default value: 25. \n",
      "\n",
      "--smith-waterman-dangling-end-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for dangling-end recovery.  Default value: -50. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -11. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -260. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-match-value <Integer>\n",
      "                              Smith-Waterman match value for haplotype-to-reference alignment.  Default value: 200. \n",
      "\n",
      "--smith-waterman-haplotype-to-reference-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for haplotype-to-reference alignment.  Default value:\n",
      "                              -150. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-gap-extend-penalty <Integer>\n",
      "                              Smith-Waterman gap-extend penalty for read-to-haplotype alignment.  Default value: -5. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-gap-open-penalty <Integer>\n",
      "                              Smith-Waterman gap-open penalty for read-to-haplotype alignment.  Default value: -30. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-match-value <Integer>\n",
      "                              Smith-Waterman match value for read-to-haplotype alignment.  Default value: 10. \n",
      "\n",
      "--smith-waterman-read-to-haplotype-mismatch-penalty <Integer>\n",
      "                              Smith-Waterman mismatch penalty for read-to-haplotype alignment.  Default value: -15. \n",
      "\n",
      "--soft-clip-low-quality-ends <Boolean>\n",
      "                              If enabled will preserve low-quality read ends as softclips (used for DRAGEN-GATK BQD\n",
      "                              genotyper model)  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--transform-dragen-mapping-quality <Boolean>\n",
      "                              If enabled this argument will map DRAGEN aligner aligned reads with mapping quality <=250\n",
      "                              to scale up to MQ 50  Default value: false. Possible values: {true, false} \n",
      "\n",
      "--use-filtered-reads-for-annotations <Boolean>\n",
      "                              Use the contamination-filtered read maps for the purposes of annotating variants  Default\n",
      "                              value: false. Possible values: {true, false} \n",
      "\n",
      "Conditional Arguments for annotation:\n",
      "\n",
      "Valid only if \"RMSMappingQuality\" is specified:\n",
      "--allow-old-rms-mapping-quality-annotation-data <Boolean>\n",
      "                              Override to allow old RMSMappingQuality annotated VCFs to function  Default value: false.\n",
      "                              Possible values: {true, false} \n",
      "\n",
      "Conditional Arguments for readFilter:\n",
      "\n",
      "Valid only if \"MappingQualityReadFilter\" is specified:\n",
      "--maximum-mapping-quality <Integer>\n",
      "                              Maximum mapping quality to keep (inclusive)  Default value: null. \n",
      "\n",
      "--minimum-mapping-quality <Integer>\n",
      "                              Minimum mapping quality to keep (inclusive)  Default value: 20. \n",
      "\n",
      "\n",
      "***********************************************************************\n",
      "\n",
      "A USER ERROR has occurred: Argument output was missing: Argument 'output' is required\n",
      "\n",
      "***********************************************************************\n",
      "Set the system property GATK_STACKTRACE_ON_USER_EXCEPTION (--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true') to print the stack trace.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='gatk HaplotypeCaller', returncode=1)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run(\"gatk HaplotypeCaller\", shell=True, check=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5da4e259",
   "metadata": {},
   "source": [
    "Optional: run bcftools / HaplotypeCaller / freebayes on data set or at\n",
    "least show commands how files were generated\n",
    "\n",
    "# References\n",
    "\n",
    "https://link.springer.com/protocol/10.1007/978-1-4939-3578-9_11"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.12,
    "jupytext_version": "1.9.1"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "pgip"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "source_map": [
   12,
   47,
   53,
   56
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}